jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/proxy/WebSocketProxyTest.java

```diff
@@ -0,0 +1,379 @@
+
+        try (StacklessLogging stacklessLogging = new StacklessLogging(WebSocketChannel.class))
+        {
+            CompletableFuture<CoreSession> response = _client.connect(clientHandler, new URI("ws://localhost:8080/proxy/"));
+            response.get(5, TimeUnit.SECONDS);
+            clientHandler.awaitClose();
+            serverFrameHandler.awaitClose();
+            awaitProxyClose(proxyClientSide, proxyServerSide);
+        }
+
+        CloseStatus closeStatus = CloseStatus.getCloseStatus(proxyClientSide.receivedFrames.poll());
+        assertThat(closeStatus.getCode(), is(CloseStatus.SERVER_ERROR));
+        assertThat(closeStatus.getReason(), containsString("simulated client onOpen error"));
+        assertThat(proxyClientSide.getState(), is(WebSocketProxy.State.CLOSED));
+
+        closeStatus = CloseStatus.getCloseStatus(proxyServerSide.receivedFrames.poll());
+        assertThat(closeStatus.getCode(), is(CloseStatus.SERVER_ERROR));
+        assertThat(closeStatus.getReason(), containsString("simulated client onOpen error"));
+        assertThat(proxyServerSide.getState(), is(WebSocketProxy.State.CLOSED));
+
+        closeStatus = CloseStatus.getCloseStatus(serverFrameHandler.receivedFrames.poll());
+        assertThat(closeStatus.getCode(), is(CloseStatus.SERVER_ERROR));
+        assertThat(closeStatus.getReason(), containsString("simulated client onOpen error"));
+
+        assertNull(clientHandler.receivedFrames.poll());
+    }
+
+
+
+    @Test
+    public void testServerError() throws Exception
+    {
+        serverFrameHandler.throwOnFrame();
+        WebSocketProxy.Client2Proxy proxyClientSide = proxy.client2Proxy;
+        WebSocketProxy.Server2Proxy proxyServerSide = proxy.server2Proxy;
+
+        BasicFrameHandler clientHandler = new BasicFrameHandler("CLIENT");
+        ClientUpgradeRequest upgradeRequest = ClientUpgradeRequest.from(_client, new URI("ws://localhost:8080/proxy/test"), clientHandler);
+
+        CompletableFuture<CoreSession> response = _client.connect(upgradeRequest);
+        response.get(5, TimeUnit.SECONDS);
+        clientHandler.sendText("hello world");
+        clientHandler.awaitClose();
+        serverFrameHandler.awaitClose();
+        awaitProxyClose(proxyClientSide, proxyServerSide);
+
+        CloseStatus closeStatus;
+        Frame frame;
+
+        // Client
+        frame = clientHandler.receivedFrames.poll();
+        assertThat(CloseStatus.getCloseStatus(frame).getCode(), is(CloseStatus.SERVER_ERROR));
+
+        // Client2Proxy
+        frame = proxyClientSide.receivedFrames.poll();
+        assertThat(frame.getOpCode(), is(OpCode.TEXT));
+        assertThat(frame.getPayloadAsUTF8(), is("hello world"));
+
+        frame = proxyClientSide.receivedFrames.poll();
+        closeStatus = CloseStatus.getCloseStatus(frame);
+        assertThat(closeStatus.getCode(), is(CloseStatus.SERVER_ERROR));
+        assertThat(closeStatus.getReason(), is("intentionally throwing in server onFrame()"));
```

Comment:
This is weird. `proxyClientSide.receivedFrames` contains the frames that the client sent to the proxy. Why does it have a close frame with `SERVER_ERROR`?
I would have expected that to be in `proxyServerSide.receivedFrames`.
If this is the reply to the close frame sent by the server, then you should reorganize the test to do the assertions in order, with a bit of comment on what should happen.

Reply:
I'm guessing this is because there is no good client side equivalent to `SERVER_ERROR`.  ie there is no `CLIENT_ERROR`.